admin/server/auth/claims.go

```diff
@@ -75,3 +85,73 @@ func (c *authTokenClaims) OwnerID() string {
 func (c *authTokenClaims) AuthTokenID() string {
 	return c.token.Token().ID.String()
 }
+
+func (c *authTokenClaims) CanOrg(orgID string, p OrganizationPermission) bool {
+	t := c.token.Token().Type
+	switch t {
+	case authtoken.TypeUser:
+		role, err := c.service.DB.ResolveUserOrganizationRole(context.Background(), c.token.OwnerID(), orgID)
+		if err != nil {
+			// TODO: log error
+			return false
+		}
+		switch p {
+		case ReadOrg:
+			return role.ReadOrganization
+		case ManageOrg:
+			return role.ManageOrganization
+		case ReadProjects:
+			return role.ReadProjects
+		case CreateProjects:
+			return role.CreateProjects
+		case ManageProjects:
+			return role.ManageProjects
+		case ReadOrgMembers:
+			return role.ReadOrgMembers
+		case ManageOrgMembers:
+			return role.ManageOrgMembers
+		default:
+			// TODO: log error
+			return false
+		}
+	default:
+		// TODO: log error
+		return false
+	}
+}
+
+func (c *authTokenClaims) CanProject(projectID string, p ProjectPermission) bool {
+	t := c.token.Token().Type
+	switch t {
+	case authtoken.TypeUser:
+		role, err := c.service.DB.ResolveUserProjectRole(context.Background(), c.token.OwnerID(), projectID)
+		if err != nil {
+			// TODO: log error
+			return false
+		}
+		switch p {
+		case ReadProject:
+			return role.ReadProject
+		case ManageProject:
+			return role.ManageProject
+		case ReadProdBranch:
+			return role.ReadProdBranch
+		case ManageProdBranch:
+			return role.ManageProdBranch
+		case ReadDevBranches:
+			return role.ReadDevBranches
+		case ManageDevBranches:
+			return role.ManageDevBranches
+		case ReadProjectMembers:
+			return role.ReadProjectMembers
+		case ManageProjectMembers:
+			return role.ManageProjectMembers
+		default:
+			// TODO: log error
+			return false
+		}
+	default:
+		// TODO: log error
+		return false
```

Comment:
for the second case added this 
```
	case authtoken.TypeService:
		return false
	default:
		panic(fmt.Errorf("unexpected token type %q", t))
```
as panicking for service token will cause application to exit in case svc token is sent currently we don't support it but just denying service instead of panicking seems better. What do you think ?

Another question I had was while checking claims we currently do this -
```
	// Check the request is made by a user
	claims := auth.GetClaims(ctx)
	if claims.OwnerType() != auth.OwnerTypeUser {
		return nil, status.Error(codes.Unauthenticated, "not authenticated")
	}
```
should we remove the `OwnerType` check as currently for other owner type `anon` we already return false from `CanXXX` methods. What can be other owner types - super user ?

Reply:
1. We have not currently implemented service tokens, so if we receive those and have not handled that case, it's a developer error, so we want to `panic`. (In prod, we do catch panics for API calls and return 500 errors.) Panics should be used to signal "the developer messed up here", so I think it fits here.
2. You're right that generally we should not need to check the `OwnerType`, rather we just call `Can..` to check permissions directly. However, there are a few places where we check the owner type because we want to be sure only a *user* can call it (i.e. because the code assumes that `OwnerID` is a valid `UserID`).